coordinate-systems.md

```diff
@@ -0,0 +1,246 @@
+# WebXR - XRCoordinateSystems et al.
+
+# Overview
+
+In systems that deal with 3D spaces and the placement, rotation, and relationships between objects within those spaces, coordinate systems are a standard way of organizing those relationships. In WebXR, there are two main classes which handle this functionality - XRCoordinateSystem and XRFrameOfReference.
+
+Currently, XRFrameOfReference refers to an enumerated set of standard coordinate systems relative to which all others must be able to be expressed. These include:
+
+
+
+*   _eye-level_: the origin and rotation of this frame of reference are the orientation of the device at the start of session.
+*   _stage_: the origin and rotation of this frame of reference are in the center of a known safe playing area like a VR room.
+*   _head-model_: this frame of reference is like eye-level except the translation component of the device pose is basically identity with a neck model. This is for experiences which expect the user to not move and only look around.
+
+These are basically different ways of expressing the transform of the device in 3D space. For example, virtual objects would have the same transform in _eye-level_, _stage_, and _head-model_ FORs, it's only the device that behaves differently (more on this later). _Stage_ basically allows for a virtual translation to be applied to the native device transform and _head-model_ always locks the device translation to (0, 0, 0).
```

Comment:
I don't agree with this point - the virtual objects have the same transforms across both of these contexts, it is only the device transform (and thus the virtual camera transform) that changes. To try to change the whole world to handle different styles of representing the device seems roundabout and involves more math on the part of the API/App.

Reply:
Indeed - for non-world-scale apps, the app is likely to parent the device-controlled camera under a camera "rig" object that they move explicitly as the user teleports or otherwise navigates around a static scene.  In that case, the engine's world origin for the app's scene is fully arbitrary and does not map to any particular physical location in the user's surroundings.

World-scale apps, in contrast, tend to not have a pre-baked static world that the user is navigating, since the device will discover the nature of the user's world geometry dynamically at runtime.  In that case, there tends not to be a reason for the app to teleport or otherwise navigate the camera's parent "rig" object, since the user would immediately notice any separation between the motion of the physical world and the virtual objects around them.

Perhaps the semantic confusion here is in saying that "virtual objects would have the same transform in" various FORs - as phrased, a virtual object's transform would necessarily differ based on the FOR used for comparison.

For example, an app may reason about "where a virtual lever in my static scene is relative to the stage origin", for which there is a concrete answer the app could compute (e.g. (10.0, 1.6, 2.0)).  This would, for example, let the app reason about whether the lever is within reach of the user within the stage's bounds (e.g. 5m x 4m) without the user having to teleport within the virtual world (e.g. no, the lever coordinates above are outside those bounds).  If not, the app may wish to relayout its virtual content.  That virtual object position would differ relative to the lever's "transform in" the eye-level FOR.

If the display device was the only thing the app ever needed to locate relative to an FOR, I could see WebXR canonizing one particular FOR (defined strictly across UAs) as the implicit bedrock frame in which all other poses are implicitly communicated.  However, given that we will also be communicating the ever-adjusting locations of things like planes, meshes and app-created free-space anchors, it's not clear what single bedrock frame we could choose that would make all apps happy:
* If we chose a user-stabilized `"world"` frame in which to express all device/controller/anchor/etc. poses, this would add extra math for room-scale `"stage"` apps that want to directly reason about the head and controller's offset relative to the virtual floor point to which the user has teleported.
* If we chose a calibrated-room `"stage"` frame in which to express all device/controller/anchor/etc. poses, we'd create confusing edge cases for world-scale apps to work through when reasoning about plane/mesh poses once the device regains tracking after moving through dark hallways (perhaps in the same map fragment as the stage, perhaps not) or if a stage turns out not to have been set up by the user in the first place.

Given that users will generally be using a middleware layer above WebXR (perhaps geared more for room-scale or world-scale apps) that establishes the conventions around the app's scene graph, I'm inclined to leave this opinionated decision about bedrock frames to that middleware layer.  By making frames of reference "just another coordinate system" and letting apps transform between any two coordinate systems as needed, we give WebXR-backed middleware such as A-Frame, Babylon.js, Unity or Sumerian the tools to build as constrained or as open-ended an abstraction as they like for their own developers.